### ISSUE TITLE
Problems with SystemBrowser API

### ISSUE THREAD

**Issue by mrueger42** [guillermoamaral/Webside#7]
Having problems with making progress with the SystemBrowser. 
Selecting from the list of projects/categories I get 
"TypeError: items.map is not a function"

Using Pharo is does nothing or shows a white screen as #projectClasses is missing in the Pharo API implementation.

Can't figure out what is supposed to be returned for e.g.
http://localhost:9001/pharo/projects/Alien-Core/classes?tree=true

**Comment by guillermoamaral** [comment]
Hi there!

Thanks for reporting this. You've reached some of the not-well defined points: it turns out that the different dialects manage code bundles differently. For instance, in Pharo you have packages, and within them you have tags (most of them arbitrary named but with one for [Your package] Extensions, named appropriately for that purpose). In Bee Smalltalk you have projects organized in an arbitrary deep hierarchy, where leaves are the actual projects while projects containing projects (not leaves) are called clusters. The list of different approaches continues if you go to Dolphin, VAST, etc.
So, I left that part of both the API and the front-end quite undefined. Notice that I named the tool System Browser following the Pharo nomenclature but then I didn't continue with that tool until I had a better and common definition of code modules. That is why the Pharo implementation of the API is not complete as well as the API documentation in Github. Regarding your specific question about what the API should return when sending a request to http://localhost:9001/pharo/projects/Alien-Core/classes?tree=true is the classes in a tree structure (like the one you get when you use http://localhost:9001/pharo/classes?root=Magnitude&tree=true, for instance), included in the Alien-Core package.
Note also (I'm remembering while I write) that compiled methods at the level of the API (JSON) include a property "project" that refers (and it's used to filter them) the very same thing that I described above: the module where the method is defined (in the case of a Pharo package like Alien-Core, methods coming from that packages should include a property "project": "Alien-Core").

Thanks again for your feedback. I will update the documentation, fix the front-end to avoid the error and (maybe) complete the Pharo implementation of the API.

Regards,
________________________________

Guillermo Amaral
Development Manager

t: +1 281 254 7600
m: +54 9 11 63 03 76 82
Confidentiality Notice:
 The information transmitted herein may contain confidential, proprietary and/or privileged material which belongs to Aucerna and is only for the intended addressee(s). Any unauthorized review, distribution or other use of this information is prohibited. Finally, the recipient should check this email and any attachments for the presence of viruses. The organization accepts no liability for any damage caused by any virus transmitted by this email. If you received this email in error, please contact us at (403) 270-3270 or by return email and delete this email and any copies immediately.  v&7.11.1989&$
From: mrueger42 ***@***.***>
Sent: Monday, May 3, 2021 2:29 AM
To: guillermoamaral/Webside ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [guillermoamaral/Webside] Problems with SystemBrowser API (#7)


ATTENTION: This email originated from outside of the organization. Do not open attachments or click any links unless you are sure this email comes from a known sender and you know the content is safe. Do not send any unsecured personal or financial information without verification.

Having problems with making progress with the SystemBrowser.
Selecting from the list of projects/categories I get
"TypeError: items.map is not a function"

Using Pharo is does nothing or shows a white screen as #projectClasses is missing in the Pharo API implementation.

Can't figure out what is supposed to be returned for e.g.
http://localhost:9001/pharo/projects/Alien-Core/classes?tree=true

**Comment by mrueger42** [comment]
thank you!
What I find most useful is to have JSON examples that cover the basic scenarios, then I can figure out how to produce that format in whatever system I'm working with.
I should try and get Bee running with Webside, that should help too :-)

**Comment by guillermoamaral** [comment]
You're welcome.

I've made work the basic endpoints in Python a few days ago... just for fun! I'm closing a version and making the repo open-source afterwards (yes... it sound strange and leaves a whole part of Python stuff outside (not pure oop), but it was fun to have the Webside debugger working!


Get Outlook para Android<https://aka.ms/AAb9ysg>
________________________________

Guillermo Amaral
Development Manager

t: +1 281 254 7600
m: +54 9 11 63 03 76 82
Confidentiality Notice:
 The information transmitted herein may contain confidential, proprietary and/or privileged material which belongs to Aucerna and is only for the intended addressee(s). Any unauthorized review, distribution or other use of this information is prohibited. Finally, the recipient should check this email and any attachments for the presence of viruses. The organization accepts no liability for any damage caused by any virus transmitted by this email. If you received this email in error, please contact us at (403) 270-3270 or by return email and delete this email and any copies immediately.  v&7.11.1989&$

**Comment by guillermoamaral** [comment]
There you go! I just pushed the implementation of #projectClasses (and a private service supporting it) to Webside-Pharo repo (https://github.com/guillermoamaral/Pharo-Webside). I tested it and it works fine except when you click on a tag (it doesn't filter the classes under it). I should see that but it's too late here in Buenos Aires.
While testing it I found another issue in the way a package was jsonified so I fixed that too.

Regards,
________________________________

Guillermo Amaral
Development Manager

t: +1 281 254 7600
m: +54 9 11 63 03 76 82
Confidentiality Notice:
 The information transmitted herein may contain confidential, proprietary and/or privileged material which belongs to Aucerna and is only for the intended addressee(s). Any unauthorized review, distribution or other use of this information is prohibited. Finally, the recipient should check this email and any attachments for the presence of viruses. The organization accepts no liability for any damage caused by any virus transmitted by this email. If you received this email in error, please contact us at (403) 270-3270 or by return email and delete this email and any copies immediately.  v&7.11.1989&$

**Comment by mrueger42** [comment]
Cheers, appreciate it!

**Comment by guillermoamaral** [comment]
I appreciate any issue report, enhancement requirement or feedback coming from using Webside.

**Comment by guillermoamaral** [comment]
This doesn't apply anymore as the old SystemBrowser was replaced by the PackageBrowser.
